src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/OOPTagHelperResolver.cs

```diff
@@ -104,24 +104,21 @@ protected virtual async Task<TagHelperResolutionResult> ResolveTagHelpersOutOfPr
             // This will change in the future to an easier to consume API but for VS RTM this is what we have.
             try
             {
-                var client = await RazorLanguageServiceClientFactory.CreateAsync(_workspace, CancellationToken.None).ConfigureAwait(false);
-                if (client != null)
+                var remoteClient = await RazorRemoteHostClient.CreateAsync(workspaceProject.Solution.Workspace, CancellationToken.None);
+
+                var args = new object[]
                 {
-                    using (var session = await client.CreateSessionAsync(workspaceProject.Solution).ConfigureAwait(false))
-                    {
-                        if (session != null)
-                        {
-                            var args = new object[]
-                            {
-                                Serialize(projectSnapshot),
-                                factory == null ? null : factory.GetType().AssemblyQualifiedName,
-                            };
-
-                            var json = await session.InvokeAsync<JObject>("GetTagHelpersAsync", args, CancellationToken.None).ConfigureAwait(false);
-                            return Deserialize(json);
-                        }
-                    }
-                }
+                    Serialize(projectSnapshot),
+                    factory == null ? null : factory.GetType().AssemblyQualifiedName,
+                };
+
+                var resolutionResult = await remoteClient.TryRunRemoteAsync<TagHelperResolutionResult>(
+                    "GetTagHelpersAsync",
+                    workspaceProject.Solution,
+                    args,
+                    CancellationToken.None).ConfigureAwait(false);
+
+                return resolutionResult.Value;
```

Comment:
>resolutionResult [](start = 23, length = 16)

TryRunRemoteAsync may fail and return no value. Rather than throwing an exception this should check `if (!resolutionResult.HasValue) return null;` (this is equivalent to `session != null` check that was there before.

Reply:
I considered doing that, but when I examined the documentation for `Optional.Value` I saw:
```
// Returns:
//     The value if Microsoft.CodeAnalysis.Optional`1.HasValue is true; otherwise, the
//     default value for type T.
//
// Remarks:
//     Unlike System.Nullable`1.Value, this property does not throw an exception when
//     Microsoft.CodeAnalysis.Optional`1.HasValue is false.
```

Since the default value of `TagHelperResolutionResult` would be `null` I believe the two should be equivalent. I can add a clarifying comment since this does create some confusion with `System.Nullable`.

Reply:
Oh, ok :)